src/index.ts

```diff
@@ -0,0 +1,285 @@
+    const createUpdateQuery = (user: any) => {
+      const fields = Object.keys(user);
+      const values = Object.values(user);
+      let query = `UPDATE users SET `;
+      for (let i = 0; i < fields.length; i++) {
+        if (i === fields.length - 1) {
+          query += `${fields[i]} = $${i + 1} `;
+        } else {
+          query += `${fields[i]} = $${i + 1}, `;
+        }
+      }
+      query += `WHERE id = ${user.id} RETURNING id, email, userName, firstName, lastName;`;
+      return query;
+    }
+
+    const updatedUserResult = await db.query(createUpdateQuery(user), Object.values(user));
+    return updatedUserResult.rows[0];
+  }
+
+  public async getUserByEmail(email: string) {
+    const foundUser = await db.query(`
+    SELECT id, email, userName, firstName, lastName
+    FROM users
+    WHERE email = $1;
+    `, [email.toString()]
+    );
+    if (foundUser.rows.length === 0) {
+      return false;
+    }
+    return foundUser.rows[0];
+  }
+}
+
+/**--------------App--------------**/
+
+const userService = new UserService();
+const app = express();
+app.use(express.json());
+app.use(cors());
+
+/**--------------Routes--------------**/
+
+// Create a new user
+app.post('/users/new', async (req: Request, res: Response) => {
+  const user = req.body;
+  const newUser = {
+    ...user,
+    password: generatePassword(),
+  }
+
+  if (!newUser.email || !newUser.userName) {
+    res.status(400).json({
+      error: 'ValidationError',
+      data: undefined,
+      success: false,
+    });
+    return;
+  }
+
+  try {
+    const userNameTaken = await userService.userNameTaken(newUser.userName);
+    if (userNameTaken) {
+     return res.status(409).json({
+        error: 'UsernameAlreadyTaken',
+        data: undefined,
+        success: false,
+      });
+    }
+
+    const emailAlreadyInUse = await userService.emailAlreadyInUse(newUser.email);
+    if (emailAlreadyInUse) {
+      return res.status(409).json({
+        error: 'EmailAlreadyInUse',
+        data: undefined,
+        success: false,
+      });
+    }
+
+    const user = await userService.createUser(newUser);
+    res.status(201).json({
+      error: undefined,
+      data: {
+        id: user.id,
+        email: user.email,
+        userName: user.userName,
+        firstName: user.firstName,
+        lastName: user.lastName,
+      },
+      success: true,
+    });
+  } catch (error: unknown) {
+    return res.status(500).json({
+      error: 'ServerError',
+      data: undefined,
+      success: false,
+    });
+  }
+});
+
+app.post('/users/edit/:userId', async (req: Request, res: Response) => {
+  const userId = req.params.userId;
+  const user = req.body;
+
+  if (!user.email && !user.userName && !user.firstName && !user.lastName) {
+    return res.status(400).json({
+      error: 'ValidationError',
+      data: undefined,
+      success: false,
+    });
+  }
```

Comment:
I wasn't really sure on the assignment on how to handle this. Cause technically, only `userId` and at least one of the fields is required for an update. The others could technically be left out.

Just thought I'd point this out. But it's not really an issue if you know the FE will send the whole user through.